File: src/commands/fs.ts
import { invoke } from "@tauri-apps/api/core"
import type { FileNode, WikiProject } from "@/types/wiki"
import { ensureProjectId, upsertProjectInfo } from "@/lib/project-identity"
import { isTauri } from "@/lib/platform"
import { getWebFs } from "@/lib/web-fs"

interface RawProject {
  name: string
  path: string
}

export async function readFile(path: string): Promise<string> {
  if (!isTauri()) {
    return getWebFs().readFile(path)
  }
  return invoke<string>("read_file", { path })
}

export async function writeFile(path: string, contents: string): Promise<void> {
  if (!isTauri()) {
    return getWebFs().writeFile(path, contents)
  }
  return invoke<void>("write_file", { path, contents })
}

export async function writeFileAtomic(path: string, contents: string): Promise<void> {
  if (!isTauri()) {
    return getWebFs().writeFileAtomic(path, contents)
  }
  return invoke<void>("write_file_atomic", { path, contents })
}

export async function listDirectory(path: string): Promise<FileNode[]> {
  if (!isTauri()) {
    return getWebFs().listDirectory(path)
  }
  return invoke<FileNode[]>("list_directory", { path })
}

export async function copyFile(
  source: string,
  destination: string
): Promise<void> {
  if (!isTauri()) {
    return getWebFs().copyFile(source, destination)
  }
  return invoke("copy_file", { source, destination })
}

export async function copyDirectory(
  source: string,
  destination: string
): Promise<string[]> {
  if (!isTauri()) {
    return getWebFs().copyDirectory(source, destination)
  }
  return invoke<string[]>("copy_directory", { source, destination })
}

export async function preprocessFile(path: string): Promise<string> {
  if (!isTauri()) {
    return getWebFs().preprocessFile(path)
  }
  return invoke<string>("preprocess_file", { path })
}

export async function deleteFile(path: string): Promise<void> {
  if (!isTauri()) {
    return getWebFs().deleteFile(path)
  }
  return invoke("delete_file", { path })
}

export async function findRelatedWikiPages(
  projectPath: string,
  sourceName: string
): Promise<string[]> {
  if (!isTauri()) {
    return getWebFs().findRelatedWikiPages(projectPath, sourceName)
  }
  return invoke<string[]>("find_related_wiki_pages", { projectPath, sourceName })
}

export async function createDirectory(path: string): Promise<void> {
  if (!isTauri()) {
    return getWebFs().createDirectory(path)
  }
  return invoke<void>("create_directory", { path })
}

export async function fileExists(path: string): Promise<boolean> {
  if (!isTauri()) {
    return getWebFs().fileExists(path)
  }
  return invoke<boolean>("file_exists", { path })
}

export async function getFileModifiedTime(path: string): Promise<number> {
  if (!isTauri()) {
    return getWebFs().getFileModifiedTime(path)
  }
  return invoke<number>("get_file_modified_time", { path })
}

export async function getFileSize(path: string): Promise<number> {
  if (!isTauri()) {
    return getWebFs().getFileSize(path)
  }
  return invoke<number>("get_file_size", { path })
}

export async function getFileMd5(path: string): Promise<string> {
  if (!isTauri()) {
    return getWebFs().getFileMd5(path)
  }
  return invoke<string>("get_file_md5", { path })
}

export interface FileBase64 {
  base64: string
  mimeType: string
}

export async function readFileAsBase64(path: string): Promise<FileBase64> {
  if (!isTauri()) {
    return getWebFs().readFileAsBase64(path)
  }
  return invoke<FileBase64>("read_file_as_base64", { path })
}

export async function createProject(
  name: string,
  path: string,
): Promise<WikiProject> {
  if (!isTauri()) {
    const fs = getWebFs()
    const raw = await fs.createProject(name, path)
    const id = `web-${Date.now()}`
    await upsertProjectInfo(id, raw.path, raw.name)
    return { id, name: raw.name, path: raw.path }
  }
  const raw = await invoke<RawProject>("create_project", { name, path })
  const id = await ensureProjectId(raw.path)
  await upsertProjectInfo(id, raw.path, raw.name)
  return { id, name: raw.name, path: raw.path }
}

export async function openProject(path: string): Promise<WikiProject> {
  if (!isTauri()) {
    const fs = getWebFs()
    const raw = await fs.openProject(path)
    const id = `web-${Date.now()}`
    await upsertProjectInfo(id, raw.path, raw.name)
    return { id, name: raw.name, path: raw.path }
  }
  const raw = await invoke<RawProject>("open_project", { path })
  const id = await ensureProjectId(raw.path)
  await upsertProjectInfo(id, raw.path, raw.name)
  return { id, name: raw.name, path: raw.path }
}

export async function openProjectFolder(path: string): Promise<void> {
  if (!isTauri()) {
    return getWebFs().openProjectFolder(path)
  }
  return invoke<void>("open_project_folder", { path })
}

export async function openFileLocation(path: string): Promise<void> {
  if (!isTauri()) {
    const { httpProject } = await import("@/lib/http-adapter")
    return httpProject.openFileLocation(path)
  }
  return invoke<void>("open_file_location", { path })
}

export async function clipServerStatus(): Promise<string> {
  if (!isTauri()) {
    return getWebFs().clipServerStatus()
  }
  return invoke<string>("clip_server_status")
}

export async function getExecutableDir(): Promise<string> {
  if (!isTauri()) {
    return getWebFs().getExecutableDir?.() || process.cwd() || ""
  }
  return invoke<string>("get_executable_dir")
}

export async function getResourceDir(): Promise<string> {
  if (!isTauri()) {
    return getWebFs().getResourceDir?.() || process.cwd() || ""
  }
  return invoke<string>("get_resource_dir")
}

File: src-tauri/src/server/handlers/fs.rs
use axum::extract::Multipart;
use axum::Json;
use serde::Deserialize;

use crate::commands::fs;

// ── Request types ──────────────────────────────────────────────────

#[derive(Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PathReq {
    pub path: String,
}

#[derive(Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct WriteFileReq {
    pub path: String,
    pub contents: String,
}

#[derive(Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CopyReq {
    pub source: String,
    pub destination: String,
}

#[derive(Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct FindRelatedReq {
    pub project_path: String,
    pub source_name: String,
}

// ── Helper ─────────────────────────────────────────────────────────

fn ok<T: serde::Serialize>(data: T) -> Json<serde_json::Value> {
    Json(serde_json::json!({"ok": true, "data": data}))
}

fn err(e: String) -> Json<serde_json::Value> {
    Json(serde_json::json!({"ok": false, "error": e}))
}

// ── Handlers ───────────────────────────────────────────────────────

pub async fn read_file(Json(req): Json<PathReq>) -> Json<serde_json::Value> {
    match tokio::task::spawn_blocking(move || fs::do_read_file(&req.path)).await {
        Ok(Ok(content)) => ok(content),
        Ok(Err(e)) => err(e),
        Err(e) => err(format!("read_file join error: {e}")),
    }
}

pub async fn write_file(Json(req): Json<WriteFileReq>) -> Json<serde_json::Value> {
    match tokio::task::spawn_blocking(move || fs::do_write_file(&req.path, &req.contents)).await {
        Ok(Ok(_)) => ok(serde_json::Value::Null),
        Ok(Err(e)) => err(e),
        Err(e) => err(format!("write_file join error: {e}")),
    }
}

pub async fn write_file_atomic(Json(req): Json<WriteFileReq>) -> Json<serde_json::Value> {
    match tokio::task::spawn_blocking(move || fs::do_write_file_atomic(&req.path, &req.contents))
        .await
    {
        Ok(Ok(_)) => ok(serde_json::Value::Null),
        Ok(Err(e)) => err(e),
        Err(e) => err(format!("write_file_atomic join error: {e}")),
    }
}

pub async fn list_directory(Json(req): Json<PathReq>) -> Json<serde_json::Value> {
    match tokio::task::spawn_blocking(move || fs::do_list_directory(&req.path)).await {
        Ok(Ok(nodes)) => ok(nodes),
        Ok(Err(e)) => err(e),
        Err(e) => err(format!("list_directory join error: {e}")),
    }
}

pub async fn copy_file(Json(req): Json<CopyReq>) -> Json<serde_json::Value> {
    match tokio::task::spawn_blocking(move || fs::do_copy_file(&req.source, &req.destination)).await
    {
        Ok(Ok(_)) => ok(serde_json::Value::Null),
        Ok(Err(e)) => err(e),
        Err(e) => err(format!("copy_file join error: {e}")),
    }
}

pub async fn copy_directory(Json(req): Json<CopyReq>) -> Json<serde_json::Value> {
    match tokio::task::spawn_blocking(move || fs::do_copy_directory(&req.source, &req.destination))
        .await
    {
        Ok(Ok(files)) => ok(files),
        Ok(Err(e)) => err(e),
        Err(e) => err(format!("copy_directory join error: {e}")),
    }
}

pub async fn preprocess_file(Json(req): Json<PathReq>) -> Json<serde_json::Value> {
    match tokio::task::spawn_blocking(move || fs::do_preprocess_file(&req.path)).await {
        Ok(Ok(text)) => ok(text),
        Ok(Err(e)) => err(e),
        Err(e) => err(format!("preprocess_file join error: {e}")),
    }
}

pub async fn delete_file(Json(req): Json<PathReq>) -> Json<serde_json::Value> {
    match tokio::task::spawn_blocking(move || fs::do_delete_file(&req.path)).await {
        Ok(Ok(_)) => ok(serde_json::Value::Null),
        Ok(Err(e)) => err(e),
        Err(e) => err(format!("delete_file join error: {e}")),
    }
}

pub async fn find_related_wiki_pages(Json(req): Json<FindRelatedReq>) -> Json<serde_json::Value> {
    match tokio::task::spawn_blocking(move || {
        fs::do_find_related_wiki_pages(&req.project_path, &req.source_name)
    })
    .await
    {
        Ok(Ok(pages)) => ok(pages),
        Ok(Err(e)) => err(e),
        Err(e) => err(format!("find_related_wiki_pages join error: {e}")),
    }
}

pub async fn create_directory(Json(req): Json<PathReq>) -> Json<serde_json::Value> {
    match tokio::task::spawn_blocking(move || fs::do_create_directory(&req.path)).await {
        Ok(Ok(_)) => ok(serde_json::Value::Null),
        Ok(Err(e)) => err(e),
        Err(e) => err(format!("create_directory join error: {e}")),
    }
}

pub async fn file_exists(Json(req): Json<PathReq>) -> Json<serde_json::Value> {
    match tokio::task::spawn_blocking(move || fs::do_file_exists(&req.path)).await {
        Ok(Ok(exists)) => ok(exists),
        Ok(Err(e)) => err(e),
        Err(e) => err(format!("file_exists join error: {e}")),
    }
}

pub async fn get_file_modified_time(Json(req): Json<PathReq>) -> Json<serde_json::Value> {
    match tokio::task::spawn_blocking(move || fs::do_get_file_modified_time(&req.path)).await {
        Ok(Ok(ts)) => ok(ts),
        Ok(Err(e)) => err(e),
        Err(e) => err(format!("get_file_modified_time join error: {e}")),
    }
}

pub async fn get_file_size(Json(req): Json<PathReq>) -> Json<serde_json::Value> {
    match tokio::task::spawn_blocking(move || fs::do_get_file_size(&req.path)).await {
        Ok(Ok(size)) => ok(size),
        Ok(Err(e)) => err(e),
        Err(e) => err(format!("get_file_size join error: {e}")),
    }
}

pub async fn get_file_md5(Json(req): Json<PathReq>) -> Json<serde_json::Value> {
    match tokio::task::spawn_blocking(move || fs::do_get_file_md5(&req.path)).await {
        Ok(Ok(hash)) => ok(hash),
        Ok(Err(e)) => err(e),
        Err(e) => err(format!("get_file_md5 join error: {e}")),
    }
}

pub async fn read_file_as_base64(Json(req): Json<PathReq>) -> Json<serde_json::Value> {
    match tokio::task::spawn_blocking(move || fs::do_read_file_as_base64(&req.path)).await {
        Ok(Ok(file_b64)) => ok(file_b64),
        Ok(Err(e)) => err(e),
        Err(e) => err(format!("read_file_as_base64 join error: {e}")),
    }
}

pub async fn get_executable_dir() -> Json<serde_json::Value> {
    match fs::do_get_executable_dir() {
        Ok(dir) => ok(dir),
        Err(e) => err(e),
    }
}

pub async fn get_resource_dir() -> Json<serde_json::Value> {
    match fs::do_get_resource_dir() {
        Ok(dir) => ok(dir),
        Err(e) => err(e),
    }
}

/// Multipart 文件上传（浏览器模式导入章节/大纲文件使用）
/// 接收多个文件，保存到临时目录，返回临时目录路径和文件路径列表
/// 前端 FormData 格式：
///   - paths: JSON 字符串数组，每个元素为文件的相对路径（与 file 字段一一对应）
///   - file: 文件内容（多个，字段名均为 "file"，顺序与 paths 对应）
/// 如果 paths 数量与 file 数量不匹配，则回退到使用原始文件名（平铺保存）
pub async fn upload_files(mut multipart: Multipart) -> Json<serde_json::Value> {
    let temp_dir = std::env::temp_dir().join(format!("qmai-upload-{}", uuid::Uuid::new_v4()));
    if let Err(e) = std::fs::create_dir_all(&temp_dir) {
        return err(format!("创建临时目录失败: {e}"));
    }

    let mut rel_paths: Vec<String> = Vec::new();
    let mut file_entries: Vec<(String, Vec<u8>)> = Vec::new(); // (原始文件名, 数据)

    while let Ok(Some(field)) = multipart.next_field().await {
        let field_name = field.name().unwrap_or("").to_string();

        match field_name.as_str() {
            "paths" => {
                if let Ok(text) = field.text().await {
                    let _ = serde_json::from_str::<Vec<String>>(&text).map(|p| rel_paths = p);
                }
            }
            "file" => {
                let file_name = field.file_name().unwrap_or("unnamed").to_string();
                match field.bytes().await {
                    Ok(bytes) => {
                        file_entries.push((file_name, bytes.to_vec()));
                    }
                    Err(e) => return err(format!("读取上传文件失败: {e}")),
                }
            }
            _ => {
                let _ = field.bytes().await;
            }
        }
    }

    if file_entries.is_empty() {
        let _ = std::fs::remove_dir_all(&temp_dir);
        return err("未找到上传文件".to_string());
    }

    let use_rel_paths = rel_paths.len() == file_entries.len();
    let mut saved_paths: Vec<String> = Vec::new();

    for (i, (original_name, data)) in file_entries.iter().enumerate() {
        let rel = if use_rel_paths {
            rel_paths[i].clone()
        } else {
            original_name.clone()
        };

        // 安全处理路径：去除驱动器号和前导分隔符，防止路径穿越
        let safe_rel = rel
            .replace('\\', "/")
            .split('/')
            .filter(|s| !s.is_empty() && !s.contains(':') && *s != "..")
            .collect::<Vec<_>>()
            .join("/");

        if safe_rel.is_empty() {
            continue;
        }

        let dest = temp_dir.join(&safe_rel);

        // 创建父目录
        if let Some(parent) = dest.parent() {
            let _ = std::fs::create_dir_all(parent);
        }

        if let Err(e) = std::fs::write(&dest, data) {
            return err(format!("保存文件失败: {e}"));
        }
        saved_paths.push(dest.to_string_lossy().to_string());
    }

    if saved_paths.is_empty() {
        let _ = std::fs::remove_dir_all(&temp_dir);
        return err("未找到上传文件".to_string());
    }

    ok(serde_json::json!({
        "tempDir": temp_dir.to_string_lossy(),
        "paths": saved_paths,
    }))
}
